src/Kernel/CompiledCode.class.st

```diff
@@ -295,15 +295,15 @@ CompiledCode >> hasSelector: selector specialSelectorIndex: specialOrNil [
 	 If you do, you may call this method directly to avoid recomputing 
 	 the special selector index all the time.
 	
-	 I traverse the method and all the compiled blocks in the literals 	 
-	"
-
-	^ self withAllBlocks
-		anySatisfy: [ :aCompiledCode | 
-			(aCompiledCode refersToLiteral: selector)
-				or: [ specialOrNil isNotNil
-						and: [ aCompiledCode
-								scanFor: self encoderClass firstSpecialSelectorByte + specialOrNil ] ] ]
+	 I traverse the method and all the compiled blocks in the literals"
+
+	
+	(self refersToLiteral: selector) ifTrue: [ ^ true ].
+	"refersToLiteral: traverses all blocks, but only for non-special literals"
+	specialOrNil ifNil: [ ^ false ].
+	"if the selector is special, scan all blocks and myself"
+	^ self withAllBlocks anySatisfy: [ :aCompiledCode | aCompiledCode scanFor:
```

Comment:
for more optimizations we should add #withAllBlocksDo: to avoid extra objects allocation.

Reply:
yes, another thing is that we should use the same pattern for all these methods, right now there are 3 ways that we implement a method like this... (besides this: have local* method doing it locally, then do a recursive call, third way is to  implement the method on all literals and just call it when iterating over the literals, leading to recursion on embedded blocks)